technology_and_tooling/testing/mocking.md

```diff
@@ -0,0 +1,526 @@
+        pass
+
+class MockSQLiteCursor:
+
+    def execute(self, query):
+        # Implement a method to simulate executing a query
+        pass
+
+    def fetchall(self):
+        # Implement a method to simulate fetching a result
+        return [("Jerry", "Mouse", 1)]
+
+
+def test_query_db_mocked_connection(monkeypatch):
+    # a function that returns our mock connection
+    def mock_connection(*args, **kwargs):
+        return MockSQLiteConnection(*args)
+    # replace the sqlite3 connect() method with our function
+    monkeypatch.setattr("sqlite3.connect", mock_connection)
+    # make our fake connection
+    conn = sqlite3.connect("my_non_existent_file")
+    sql = "SELECT * FROM Animals"
+    # test what query_database does with our fake connection
+    result = query_database(sql, connection=conn)
+    assert result[0] == ("Jerry", "Mouse", 1)
+
+~~~
+
+As can be seen, if we do not use the `unittest.mock` library at all, we have to write our own  `MockSQLiteConnection` and `MockSQLiteCursor` objects. Also we can no longer use the functions that allow us to check how our functions interacted with the mock objects such as `assert_called_once_with()`. In reality, it would make more sense to use both the pytest `monkeypatch` fixture and `unittest.mock`.
+
+::::challenge{id=mock-monkeypatch title="Using `monkeypatch` and `Mock`"}
+
+Rewrite `test_query_db_mocked_connection` to use the pytest `monkeypatch` fixture alongside `unittest.mock`. You can view the [pytest monkeypatch documentation](https://docs.pytest.org/en/7.1.x/how-to/monkeypatch.html) if needed.
+
+:::solution
+~~~python
+import pytest
+import sqlite3
+from sqlite_example import query_database
+from unittest.mock import patch, Mock
+
+def test_query_db_mocked_connection(monkeypatch):
+    """Mock the database connection and cursor to ensure the correct methods
+    are called within the query_database function
+    """
+    # Create a mock sqlite3.Connection object
+    mock_conn = Mock()
+
+    # Create a mock sqlite3.Cursor object
+    mock_cursor = Mock()
+    mock_cursor.fetchall.return_value = [("Jerry", "Mouse", 1)]
+    mock_conn.cursor.return_value = mock_cursor
+
+    # Monkeypatch the sqlite3.connect function with a mock
+    def mock_connect(*args, **kwargs):
+        return mock_conn
+
+    monkeypatch.setattr("sqlite3.connect", mock_connect)
+    conn = sqlite3.connect("my_non_existent_file")
+    # Call the function and assert the expected behavior
+    sql = "SELECT * FROM Animals"
+    result = query_database(sql, connection=conn)
+    assert result[0] == ("Jerry", "Mouse", 1)
+    # check that query_database passes our SQL string to cursor.execute()
+    mock_cursor.execute.assert_called_once_with(sql)
+    # check that fetchall() was called
+    mock_cursor.fetchall.assert_called_once()
+    # check that query_database closes the connection
+    conn.close.assert_called_once()
+    
+~~~
+
+:::
+::::
+
+#### Using the `mocker` fixture from `pytest-mock`
```

Comment:
You are presenting many different options here, and I think its a bit confusing. If you think `pytest-mock` is clearly better to use than `unittest.mock` + `monkeypatch`, perhaps just present this. If not, then you can just mention `pytest-mock` as an aside / callout without giving much detail.

Reply:
`pytest-mock` is now just mentioned as a callout with a link to refer to the documentation, thanks